framework/encode/vulkan_state_table_base.h

```diff
@@ -77,38 +78,30 @@ class VulkanStateTableBase
                      Wrapper*                                                    wrapper,
                      std::unordered_map<typename Wrapper::HandleType, Wrapper*>& map)
     {
-        const std::lock_guard<std::mutex> lock(mutex_);
-        const auto&                       inserted = map.insert(std::make_pair(handle, wrapper));
+        const std::unique_lock<std::shared_mutex> lock(mutex_);
+        const auto&                               inserted = map.insert(std::make_pair(handle, wrapper));
         return inserted.second;
     }
 
     template <typename Wrapper>
     bool RemoveEntry(const typename Wrapper::HandleType                          handle,
                      std::unordered_map<typename Wrapper::HandleType, Wrapper*>& map)
     {
-        const std::lock_guard<std::mutex> lock(mutex_);
+        const std::unique_lock<std::shared_mutex> lock(mutex_);
         return (map.erase(handle) != 0);
     }
 
     template <typename Wrapper>
-    Wrapper* GetWrapper(typename Wrapper::HandleType                                handle,
-                        std::unordered_map<typename Wrapper::HandleType, Wrapper*>& map)
+    Wrapper* GetWrapper(typename Wrapper::HandleType                                      handle,
```

Comment:
Is this "read access" If any thread can modify the the `Wrapper` pointed-to, and is that thread-safe if there is no further synchronisation? (maybe there is synchronisation elsewhere for the accesses to the fields of `Wrapper`s? Sorry, I should grep this but late here.

Reply:
I have restored the const and non-const methods with the change that the non-const method now takes a `const` argument for `map`--it would not be thread-safe to modify `map` while it is protected by a `shared_lock` and not a `unique_lock`.

> maybe there is synchronisation elsewhere for the accesses to the fields of Wrappers?

In general I do not think GFXR synchronizes wrapper state updates and instead relies on the application to make changes to objects in a thread-safe way. For example an application cannot record commands to a single `VkCommandBuffer` from multiple threads simultaneously and so GFXR doesn't add additional synchronization when updating tracked command buffer state in `VulkanStateTracker::TrackCommandExecution`.

Reply:
Okay, that makes sense :).